source/adapters/level_zero/context.cpp

```diff
@@ -178,111 +178,103 @@ UR_APIEXPORT ur_result_t UR_APICALL urContextSetExtendedDeleter(
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
 
+// Template helper function for creating USM pools for given pool descriptor.
+template <typename P, typename... Args>
+std::pair<ur_result_t, umf::pool_unique_handle_t>
+createUMFPoolForDesc(usm::pool_descriptor &Desc, Args &&...args) {
+  umf_result_t UmfRet = UMF_RESULT_SUCCESS;
+  umf::provider_unique_handle_t MemProvider = nullptr;
+
+  switch (Desc.type) {
+  case UR_USM_TYPE_HOST: {
+    std::tie(UmfRet, MemProvider) =
+        umf::memoryProviderMakeUnique<L0HostMemoryProvider>(Desc.hContext,
+                                                            Desc.hDevice);
+    break;
+  }
+  case UR_USM_TYPE_DEVICE: {
+    std::tie(UmfRet, MemProvider) =
+        umf::memoryProviderMakeUnique<L0DeviceMemoryProvider>(Desc.hContext,
+                                                              Desc.hDevice);
+    break;
+  }
+  case UR_USM_TYPE_SHARED: {
+    if (Desc.deviceReadOnly) {
+      std::tie(UmfRet, MemProvider) =
+          umf::memoryProviderMakeUnique<L0SharedReadOnlyMemoryProvider>(
+              Desc.hContext, Desc.hDevice);
+    } else {
+      std::tie(UmfRet, MemProvider) =
+          umf::memoryProviderMakeUnique<L0SharedMemoryProvider>(Desc.hContext,
+                                                                Desc.hDevice);
+    }
+    break;
+  }
+  default:
+    assert(0 && "Invalid UR descriptor type!");
```

Comment:
doesnt this belong to validation layer?

Reply:
No, usm::pool_descriptor is used internally.